# Проект по SQL

**Описание проекта:**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задания:**

    - Посчитайте, сколько книг вышло после 1 января 2000 года;
    - Для каждой книги посчитайте количество обзоров и среднюю оценку;
    - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа          брошюры;
    - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
    - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**План:**

    - Получаем доступ к базе данных;
    - Исследование таблиц;
    - Считаем, сколько книг вышло после 1 января 2000 года;
    - Для каждой книги считаем количество обзоров и среднюю оценку;
    - Определяем издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа            брошюры;
    - Определяем автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
    - Считаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок;
    - Общий вывод.

**Описание данных:**

Таблица books - Содержит данные о книгах:

    - book_id — идентификатор книги;
    - author_id — идентификатор автора;
    - title — название книги;
    - num_pages — количество страниц;
    - publication_date — дата публикации книги;
    - publisher_id — идентификатор издателя.
  
Таблица authors - Содержит данные об авторах:

    - author_id — идентификатор автора;
    - author — имя автора.
   
Таблица publishers - Содержит данные об издательствах:

    - publisher_id — идентификатор издательства;
    - publisher — название издательства;

Таблица ratings - Содержит данные о пользовательских оценках книг:

    - rating_id — идентификатор оценки;
    - book_id — идентификатор книги;
    - username — имя пользователя, оставившего оценку;
    - rating — оценка книги.
    
Таблица reviews - Содержит данные о пользовательских обзорах на книги:

    - review_id — идентификатор обзора;
    - book_id — идентификатор книги;
    - username — имя пользователя, написавшего обзор;
    - text — текст обзора.

## Получаем доступ к базе данных.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Исследование таблиц.
Отобразим первые строки каждой таблицы.

Таблица `books`:

In [3]:
query = '''

SELECT * 
FROM books 
LIMIT 10

'''

pd.io.sql.read_sql(sql=text(query), con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

Таблица `authors`:

In [4]:
query = '''

SELECT * 
FROM authors 
LIMIT 10

'''

pd.io.sql.read_sql(sql=text(query), con = engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

Таблица `publishers`:

In [5]:
query = '''

SELECT * 
FROM publishers 
LIMIT 10

'''

pd.io.sql.read_sql(sql=text(query), con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

Таблица `ratings`:

In [6]:
query = '''

SELECT * 
FROM ratings 
LIMIT 10

'''

pd.io.sql.read_sql(sql=text(query), con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

Таблица `reviews`:

In [7]:
query = '''

SELECT * 
FROM reviews 
LIMIT 10

'''

pd.io.sql.read_sql(sql=text(query), con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Считаем, сколько книг вышло после 1 января 2000 года.

In [8]:
query = '''

SELECT COUNT(DISTINCT title) AS amount_of_books
FROM books
WHERE CAST(publication_date AS date) > '2000-01-01' 

'''

pd.io.sql.read_sql(query, con = engine)

amount_of_books
0              818

**Вывод:**

После 1 января 2000 года вышло 818 книг.

## Для каждой книги считаем количество обзоров и среднюю оценку.

In [9]:
query = '''

SELECT b.title,
       AVG(ra.rating) AS avg_rating,
       COUNT(DISTINCT(re.review_id)) AS amount_of_reviews
FROM books AS b
LEFT JOIN ratings AS ra ON b.book_id=ra.book_id
LEFT JOIN reviews AS re ON b.book_id=re.book_id
GROUP BY b.title
            
'''

pd.io.sql.read_sql(query, con = engine)

title  avg_rating  \
0                                         'Salem's Lot    3.666667   
1                   1 000 Places to See Before You Die    2.500000   
2    13 Little Blue Envelopes (Little Blue Envelope...    4.666667   
3    1491: New Revelations of the Americas Before C...    4.500000   
4                                                 1776    4.000000   
..                                                 ...         ...   
994           Wyrd Sisters (Discworld  #6; Witches #2)    3.666667   
995                        Xenocide (Ender's Saga  #3)    3.400000   
996                                    Year of Wonders    3.200000   
997                        You Suck (A Love Story  #2)    4.500000   
998  Zen and the Art of Motorcycle Maintenance: An ...    3.833333   

     amount_of_reviews  
0                    2  
1                    1  
2                    3  
3                    2  
4                    4  
..                 ...  
994                  3  
995                  3  
996                  4  
997                  2  
998                  4  

[999 rows x 3 columns]

**Вывод:**

Для каждой книги посчтиали количество обзоров и среднюю оценку.

## Определяем издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [10]:
query = '''
            
SELECT p.publisher,
       COUNT(DISTINCT b.title) AS published_books
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY published_books DESC
LIMIT 1           
      
'''

pd.io.sql.read_sql(query, con = engine)

publisher  published_books
0  Penguin Books               42

**Вывод:**

Выяснили, что издательство `Penguin Books` выпустило 42 книги, что является наибольшим числом среди других издательств.

## Определяем автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [11]:
query = '''

WITH first AS 
    (SELECT book_id,
            AVG(rating) AS avg_rating
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating) >= 50)

SELECT author,
       AVG(avg_rating) AS avg_rating
FROM first
JOIN books AS b ON first.book_id=b.book_id
JOIN authors AS a ON b.author_id=a.author_id
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1

'''

pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

**Вывод:**

Выяснили, что у автора `J.K. Rowling/Mary GrandPré` -  самая высокой средняя оценка книг (4.28).

## Считаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [13]:
query = ''' 

WITH first AS 
    (SELECT username       
     FROM ratings
     GROUP BY username
     HAVING COUNT(rating) > 48),
    
     second AS 
    (SELECT username,
            COUNT(review_id) AS amount_of_reviews
     FROM reviews
     GROUP BY username )

SELECT AVG(amount_of_reviews) AS avg_of_reviews
FROM first
LEFT JOIN second ON second.username = first.username

'''

pd.io.sql.read_sql(query, con = engine)

avg_of_reviews
0            24.0

**Вывод:**

Вычислили, что среднее количество обзоров от пользователей, которые поставили больше 48 оценок,  24 обзора.

## Общий вывод:

В ходе проведенной работы, вывели следующие данные:

    После 1 января 2000 года вышло 818 книг.
    Для каждой книги посчтиали количество обзоров и среднюю оценку.
    Издательство Penguin Books выпустило 42 книги, что является наибольшим числом среди других издательств.
    У автора J.K. Rowling/Mary GrandPré -  самая высокой средняя оценка книг (4.28).
    Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, 24 обзора.